### PR TITLE
[ads] Transition ad history and reactions from client.json to a more efficient data approach

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/24604

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Confirm that the ad history has been successfully migrated from JSON to database.sqlite.

Confirm that the ad history is filtered appropriately, with clicks taking precedence over views and dismissals taking precedence over views.

Confirm that the ad history is sorted in descending order.

Confirm that reactions, such as liking/disliking an ad, liking/disliking a category, saving an ad, and marking an ad as inappropriate, are properly saved, restored across browser sessions, and functioning correctly.

Confirm that reminders are working as intended, specifically when a user clicks the same ad three or more times.

### CODE REVIEW

**Review by zenparsing — APPROVED** [review]

**Review by kylehickinson — APPROVED** [review]

File: browser/ui/webui/brave_rewards_page_ui.cc [hunk 1]
```diff
@@ -1115,25 +1115,28 @@ void RewardsDOMHandler::GetAdsHistory(const base::Value::List& args) {
...
-void RewardsDOMHandler::OnGetAdsHistory(base::Value::List ads_history) {
+void RewardsDOMHandler::OnGetAdsHistory(
+    std::optional<base::Value::List> ad_history) {
   if (!IsJavascriptAllowed()) {
     return;
   }
 
-  CallJavascriptFunction("brave_rewards.adsHistory", ads_history);
+  CallJavascriptFunction("brave_rewards.adsHistory", ad_history
+                                                         ? ad_history->Clone()
```
Comment (aseren) [inline browser/ui/webui/brave_rewards_page_ui.cc:1138]: it can be `std::move(*ad_history)`

Comment (tmancey) [reply browser/ui/webui/brave_rewards_page_ui.cc]: Fixed, thanks

File: components/brave_ads/core/mojom/brave_ads.mojom [hunk 2]
```diff
@@ -47,10 +47,12 @@ enum ReminderType {
   kExternalWalletConnected
 };
 
+// Do not change the order as the values are used for persisting and restoring
+// state.
 enum ReactionType {
+  kLiked = 1,
```
Comment (aseren) [inline components/brave_ads/core/mojom/brave_ads.mojom:53]: maybe put values in ascending order?

**Review by aseren — APPROVED** [review]

File: components/brave_ads/core/mojom/brave_ads.mojom (hunk 2 above)
Comment (tmancey) [reply components/brave_ads/core/mojom/brave_ads.mojom]: Fixed :-)

**Pull request merged**